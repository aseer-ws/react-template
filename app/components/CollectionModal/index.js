import { Descriptions, Modal } from 'antd';
import PropTypes from 'prop-types';
import React, { useState } from 'react';
import styled from 'styled-components';

const StyledButton = styled.button`
  background-color: transparent;
  margin-bottom: 0.5rem;
  border: none;
  cursor: pointer;
  outline: none;
  padding-left: 0;
`;

export const OK_TEXT = 'More Info';

function CollectionModal({ collectionId, collectionName, collectionPrice, collectionViewUrl, children }) {
  const [isModalVisible, setIsModalVisible] = useState(false);

  const showModal = () => {
    setIsModalVisible(true);
  };

  const handleOk = () => {
    window.open(collectionViewUrl, '_blank');
  };

  const handleCancel = () => {
    setIsModalVisible(false);
  };

  return (
    <>
      <StyledButton data-testid="modal-trigger" onClick={showModal}>
        {children}
      </StyledButton>
      <Modal
        data-testid="collection-modal"
        title="Collection Details"
        visible={isModalVisible}
        onOk={handleOk}
        okText={OK_TEXT}
        onCancel={handleCancel}
        closable
      >
        <Descriptions bordered column={2} size="small">
          <Descriptions.Item span={2} label="Collection ID">
            {collectionId}
          </Descriptions.Item>
          <Descriptions.Item span={2} label="Collection Name">
            {collectionName}
          </Descriptions.Item>
          <Descriptions.Item span={2} label="Collection Price">
            {collectionPrice} $
          </Descriptions.Item>
        </Descriptions>
      </Modal>
    </>
  );
}

CollectionModal.propTypes = {
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node]),
  collectionId: PropTypes.number,
  collectionName: PropTypes.string,
  collectionPrice: PropTypes.number,
  collectionViewUrl: PropTypes.string
};

export default CollectionModal;
