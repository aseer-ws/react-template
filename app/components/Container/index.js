import styled from 'styled-components';
import PropTypes from 'prop-types';

const Container = styled.div`
  display: flex;
  justify-content: center;
  flex-direction: column;
  margin: 2rem auto;
  max-width: ${(props) => props.maxWidth}px;
  padding: ${(props) => props.padding}rem;
`;

Container.propTypes = {
  maxWidth: PropTypes.number,
  padding: PropTypes.number
};

Container.defaultProps = {
  maxWidth: 1000,
  padding: 1
};

export default Container;
