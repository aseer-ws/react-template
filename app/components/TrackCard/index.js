import React from 'react';
import { Card, Col, Image, Row, Skeleton } from 'antd';
import { isEmpty } from 'lodash';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import If from '../If';
import { T } from '../T';
import * as colors from '@app/themes/colors';

const StyledTrackItem = styled(Col)`
  && {
    height: 100%;
    max-height: 15rem;
  }
`;

const StyledTrackCard = styled(Card)`
  && {
    height: 100%;
  }
`;

const StyledImage = styled(Image)`
  && {
    width: 6.25rem;
    height: 6.25rem;
    border-radius: 50%;

    @media screen and (max-width: 25rem) {
      width: 5rem;
      height: 5rem;
    }
  }
`;

const TrackName = styled(T)`
  && {
    font-size: 1rem;
    color: ${colors.success};
  }
`;
const TrackCollectionName = styled(T)``;
const TrackPrice = styled(T)`
  && {
    color: ${colors.error};
  }
`;
const TrackGenre = styled(T)``;

function TrackCard({ skeletonLoading, trackName, collectionName, artworkUrl100, trackPrice, primaryGenreName }) {
  return (
    <Skeleton loading={skeletonLoading} active>
      <StyledTrackItem data-testid="track-card">
        <StyledTrackCard hoverable>
          <Row gutter={48}>
            <Col span={8}>
              <StyledImage
                preview={false}
                src={artworkUrl100}
                fallback="https://media.istockphoto.com/vectors/thumbnail-image-vector-graphic-vector-id1147544807?k=20&m=1147544807&s=612x612&w=0&h=pBhz1dkwsCMq37Udtp9sfxbjaMl27JUapoyYpQm0anc="
              />
            </Col>
            <Col span={16}>
              <If condition={!isEmpty(trackName)} otherwise={<T id="track_name_unavailable" />}>
                <TrackName data-testid="track-name" marginBottom={10} type="subheading" text={trackName} />
              </If>
              <If condition={!isEmpty(collectionName)} otherwise={<T id="collection_name_unavailable" />}>
                <TrackCollectionName data-testid="collection-name" type="subText" text={collectionName} />
              </If>
              <If condition={typeof trackPrice === 'number'} otherwise={<T id="track_price_unavailable" />}>
                <TrackPrice data-testid="track-price" text={String(trackPrice)} />
              </If>
              <If condition={!isEmpty(primaryGenreName)} otherwise={<T id="track_genre_unavailable" />}>
                <TrackGenre data-testid="track-genre" text={primaryGenreName} />
              </If>
            </Col>
          </Row>
        </StyledTrackCard>
      </StyledTrackItem>
    </Skeleton>
  );
}

TrackCard.propTypes = {
  skeletonLoading: PropTypes.bool,
  trackId: PropTypes.number,
  collectionName: PropTypes.string,
  trackName: PropTypes.string,
  artworkUrl100: PropTypes.string,
  trackPrice: PropTypes.number,
  releaseDate: PropTypes.string,
  primaryGenreName: PropTypes.string
};

export default TrackCard;
