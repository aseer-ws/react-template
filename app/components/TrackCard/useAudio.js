/* eslint-disable-next-line */
import { useState, useEffect, MutableRefObject, useRef } from 'react';

/**@type { React.MutableRefObject<HTMLAudioElement> } */
let currentTrackRef;
let currentTrackId;

/**
 * @param { number } audioId
 * @param { string } audioUrl
 */
function useAudio(audioId, audioUrl) {
  const audioRef = useRef();
  const [playing, setPlaying] = useState(false);

  const togglePlay = () => setPlaying(!playing);

  useEffect(() => {
    if (playing) {
      if (audioId !== currentTrackId) {
        if (currentTrackRef?.current && !currentTrackRef.current.paused) {
          currentTrackRef.current.pause();
        }
        if (!audioRef?.current) {
          audioRef.current = new Audio(audioUrl);
          audioRef.current.id = audioId;
          audioRef.current.addEventListener('ended', function () {
            togglePlay();
            currentTrackRef = null;
            currentTrackId = null;
          });
          audioRef.current.addEventListener('pause', function () {
            if (playing) {
              togglePlay();
            }
          });
        }
      }
      currentTrackId = audioId;
      currentTrackRef = audioRef;
      audioRef?.current?.play();
    }
    if (audioRef.current) {
      if (!playing) {
        if (!audioRef.current.paused) {
          audioRef.current.pause();
        }
      }
    }
  }, [playing]);

  return {
    playing,
    togglePlay
  };
}

export default useAudio;
