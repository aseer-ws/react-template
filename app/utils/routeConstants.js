export default {
  repos: {
    route: '/repos',
    props: {
      maxwidth: 500,
      padding: 20
    },
    exact: true
  },
  track: {
    route: '/tracks/:trackId',
    props: {
      maxWidth: 800,
      padding: 1
    },
    exact: true
  },
  trackGrid: {
    route: '/',
    props: {
      maxWidth: 1500,
      padding: 1
    },
    exact: true
  },
  trackForm: {
    route: '/add-track/:stepTitle?',
    props: {
      maxWidth: 600
    },
    exact: true
  }
};
