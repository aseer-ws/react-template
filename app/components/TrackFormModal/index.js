import React from 'react';

import { Button, Modal } from 'antd';
import PropTypes from 'prop-types';
import TrackCard from '../TrackCard';

function TrackModal({ handleCancel, isVisible, trackDetails, afterClose }) {
  return (
    <Modal
      data-testid="track-modal"
      title="Submitted TrackCard"
      footer={
        <Button data-testid="close-modal" onClick={handleCancel}>
          Close
        </Button>
      }
      visible={isVisible}
      afterClose={afterClose}
      closable
    >
      <TrackCard {...trackDetails} artworkUrl100={trackDetails?.artworkUrl} />
    </Modal>
  );
}

TrackModal.propTypes = {
  isVisible: PropTypes.bool.isRequired,
  handleCancel: PropTypes.func.isRequired,
  trackDetails: PropTypes.object.isRequired,
  afterClose: PropTypes.func.isRequired
};

export default TrackModal;
