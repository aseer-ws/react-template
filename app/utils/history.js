import { createBrowserHistory } from 'history';
import { matchPath } from 'react-router';
import routeConstants from './routeConstants';

const routes = Object.keys(routeConstants);

export function findCommonRoutePrefix(routeArr) {
  let maxCommon = routeArr[0];
  for (let i = 1; i < routeArr.length - 1; i++) {
    const loopCount = routeArr[i].length > maxCommon.length ? maxCommon.length : routeArr[i + 1].length;
    stringLoop: for (let j = 0; j < loopCount; j++) {
      if (maxCommon[j] !== routeArr[i][j]) {
        maxCommon = routeArr[i + 1].substring(0, j);
        break stringLoop;
      }
    }
  }
  return maxCommon;
}

export function getBaseUrl(pathname) {
  let baseURLCollection = [];

  if (process.env.ENVIRONMENT_NAME === 'uat') {
    routes.forEach((routeKey) => {
      const route = routeConstants[routeKey].route;
      if (pathname.includes(route)) {
        // console.log('inside not : ->', { baseURL });
        if (pathname.substring(pathname.length - route.length, pathname.length) === route) {
          // console.log('matching here');
          baseURLCollection.push(pathname.substring(0, pathname.length - route.length));
        }
        if (pathname.substring(pathname.length - route.length - 1, pathname.length) === `${route}/`) {
          baseURLCollection.push(pathname.substring(0, pathname.length - route.length - 1));
        }
      } else {
        if (route.includes(':')) {
          // console.log('inside : ->', { baseURL });
          const regex = /^(?!:)\/[\w]+/;
          const matches = regex.exec(route);

          let matchLastIndex = pathname.lastIndexOf(matches[0]);
          // const matchLastIndex = pathname.lastIndexOf(matches[0]);
          const pathToMatch = pathname.substring(matchLastIndex);
          const isMatch = matchPath(pathToMatch, {
            path: route,
            exact: true
          });
          if (isMatch) {
            baseURLCollection.push(pathname.substring(0, matchLastIndex));
          }
        }
      }
    });
    if (!baseURLCollection.length) {
      // console.log('made it into undefined if section');
      return pathname;
    }
    return findCommonRoutePrefix(baseURLCollection);
  }
  return '';
}

// [DONE] http://localhost:3000/directory/ <-> '/'
// [DONE] http://localhost:3000/directory/tracks <-> '/tracks'
// [TODO] http://localhost:3000/directory <-> ''
// [DONE] http://localhost:3000/directory/tracks/123 <-> '/tracks/:trackId'
// [DONE] http://localhost:3000/directory/artist/321/tracks/123 <-> '/artist/:artistId/tracks/:trackId'
// [DONE] directory === feat/itunes
// [DONE] directory === feat/tracks

const history = createBrowserHistory({ basename: getBaseUrl(window.location.pathname) });
export default history;
