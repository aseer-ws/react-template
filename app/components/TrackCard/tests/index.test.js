import { translate } from '@app/components/IntlGlobalProvider';
import { renderProvider } from '@app/utils/testUtils';
import { fireEvent, waitFor } from '@testing-library/react';
import React from 'react';
import TrackCard, { TOGGLE_PLAY_BTN_TEST_ID } from '..';

function createSpyOnMedia(event) {
  return jest.spyOn(window.HTMLAudioElement.prototype, event).mockImplementation(() => {});
}

describe('<TrackCard /> tests', () => {
  let trackUrl;
  let playPauseButtonId;
  // let audioTrack;
  let playSpy;
  let pauseSpy;

  beforeAll(() => {
    trackUrl =
      'https://audio-ssl.itunes.apple.com/itunes-assets/AudioPreview115/v4/fb/55/23/fb552336-6b1f-99d7-31b0-55fa6f1796e1/mzaf_11062059991207198732.plus.aac.p.m4a';
    playPauseButtonId = TOGGLE_PLAY_BTN_TEST_ID;
    // audioTrack = 'audio-track';
    playSpy = createSpyOnMedia('play');
    pauseSpy = createSpyOnMedia('pause');
  });

  afterEach(() => {
    playSpy?.mockReset();
    pauseSpy?.mockReset();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderProvider(<TrackCard />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should render to the screen without props', () => {
    const { getByTestId } = renderProvider(<TrackCard />);
    expect(getByTestId('track-card')).toBeInTheDocument();
  });

  it('should render to the screen with content', () => {
    const trackName = 'Show me the meaning';
    const artworkUrl100 = 'https://testimages.org/img/testimages_screenshot.jpg';
    const collectionName = 'Backstreet Boys';
    const trackPrice = 3.1;
    const primaryGenreName = 'Pop';
    const trackCardProps = {
      trackName,
      artworkUrl100,
      collectionName,
      trackPrice,
      primaryGenreName
    };

    const { getByTestId } = renderProvider(<TrackCard {...trackCardProps} />);

    expect(getByTestId('track-name')).toHaveTextContent(trackName);
    expect(getByTestId('collection-name')).toHaveTextContent(collectionName);
    expect(getByTestId('track-price')).toHaveTextContent(trackPrice);
    expect(getByTestId('track-genre')).toHaveTextContent(primaryGenreName);
  });

  it('should render the fallback text messages if data is empty', () => {
    const { getByTestId } = renderProvider(<TrackCard />);
    expect(getByTestId('track_name_unavailable')).toHaveTextContent(translate('track_name_unavailable'));
    expect(getByTestId('collection_name_unavailable')).toHaveTextContent(translate('collection_name_unavailable'));
    expect(getByTestId('track_price_unavailable')).toHaveTextContent(translate('track_price_unavailable'));
    expect(getByTestId('track_genre_unavailable')).toHaveTextContent(translate('track_genre_unavailable'));
  });

  it('should  play the audio if clicked on play pause button', async () => {
    const { getByTestId } = renderProvider(<TrackCard previewUrl={trackUrl} />);
    fireEvent.click(getByTestId(playPauseButtonId));
    await waitFor(() => expect(playSpy).toHaveBeenCalledTimes(1));
  });

  it('should pause the playing audio if clicked on play pause button', async () => {
    const { getByTestId } = renderProvider(<TrackCard previewUrl={trackUrl} />);
    playSpy = jest.spyOn(window.HTMLAudioElement.prototype, 'play');
    pauseSpy = jest.spyOn(window.HTMLAudioElement.prototype, 'pause');
    fireEvent.click(getByTestId(playPauseButtonId));
    await waitFor(() => expect(playSpy).toHaveBeenCalledTimes(1));
    fireEvent.click(getByTestId(playPauseButtonId));
    await waitFor(() => expect(pauseSpy).toHaveBeenCalledTimes(1));
  });

  it('should set StyleImage "animate" prop to "true" when audio is playing', async () => {
    const { getByTestId } = renderProvider(<TrackCard previewUrl={trackUrl} />);
    expect(getByTestId('track-image')).toHaveAttribute('animate', 'false');
    fireEvent.click(getByTestId(playPauseButtonId));
    await waitFor(() => expect(playSpy).toHaveBeenCalledTimes(1));
    expect(getByTestId('track-image')).toHaveAttribute('animate', 'true');
  });

  // it('should prevent two audios playing at the same time', async () => {
  //   const toggleSpy = jest.fn();
  //   const { getAllByTestId } = renderProvider(
  //     <>
  //       <TrackCard trackId={123} previewUrl={trackUrl} onTrackToggle={toggleSpy} />
  //       <TrackCard trackId={321} previewUrl={trackUrl} onTrackToggle={toggleSpy} />
  //     </>
  //   );
  //   const trackElements = getAllByTestId(playPauseButtonId);
  //   playSpy = jest.spyOn(window.HTMLAudioElement.prototype, 'play');
  //   pauseSpy = jest.spyOn(window.HTMLAudioElement.prototype, 'pause');
  //   fireEvent.click(trackElements[0]);
  //   await waitFor(() => expect(playSpy).toHaveBeenCalledTimes(1));
  //   fireEvent.click(trackElements[1]);

  //   await waitFor(() => expect(toggleSpy).toHaveBeenCalledTimes(1));
  //   await waitFor(() => expect(pauseSpy).toHaveBeenCalledTimes(1));
  //   await waitFor(() => expect(playSpy).toHaveBeenCalledTimes(2));
  // });

  it('should set StyledImage prop "animate" back to "false" when playing audio is paused', async () => {
    const { getByTestId } = renderProvider(<TrackCard previewUrl={trackUrl} />);
    fireEvent.click(getByTestId(playPauseButtonId));
    await waitFor(() => expect(playSpy).toHaveBeenCalledTimes(1));
    expect(getByTestId('track-image')).toHaveAttribute('animate', 'true');
    fireEvent.pause(getByTestId('audio-track'));
    expect(getByTestId('track-image')).toHaveAttribute('animate', 'false');
  });

  // it('should set StyleImage prop "animate" back "false" when audio is ended', async () => {
  //   const { getByTestId } = renderProvider(<TrackCard previewUrl={trackUrl} />);
  //   fireEvent.click(getByTestId(playPauseButtonId));
  //   await waitFor(() => expect(playSpy).toHaveBeenCalledTimes(1));
  //   expect(getByTestId('track-image')).toHaveAttribute('animate', 'true');
  //   fireEvent.ended(getByTestId('audio-track'));
  //   expect(getByTestId('track-image')).toHaveAttribute('animate', 'false');
  // });
});
