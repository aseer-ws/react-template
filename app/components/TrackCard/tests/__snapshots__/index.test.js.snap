// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TrackCard /> tests should render and match the snapshot 1`] = `
<body>
  <div>
    <article
      class="TrackCard__StyledTrackItem-sc-10bk196-0 fmQKzS"
      data-testid="track-card"
    >
      <div
        class="ant-card ant-card-bordered TrackCard__StyledTrackCard-sc-10bk196-1 wiOID"
      >
        <div
          class="ant-card-body"
        >
          <div
            class="TrackCard__StyledImageContainer-sc-10bk196-2 bonBbJ"
          >
            <div
              animate="false"
              class="ant-image"
              data-testid="track-image"
              style="width: 100%; height: 100%;"
            >
              <img
                class="ant-image-img TrackCard__StyledImage-sc-10bk196-3 iDThZz"
                src=""
                style="height: 100%;"
              />
            </div>
          </div>
          <div
            class="TrackCard__TrackInfoContainer-sc-10bk196-5 gOyiNc"
          >
            <p
              class="T__StyledText-gjlic1-0 jbStwN TrackCard__TrackName-sc-10bk196-7 kYdkoW"
              data-testid="track_name_unavailable"
            >
              Track name is unavailable
            </p>
            <p
              class="T__StyledText-gjlic1-0 jbStwN TrackCard__TrackCollectionName-sc-10bk196-8 DicOb"
              data-testid="collection_name_unavailable"
            >
              Collection name is unavailable
            </p>
            <p
              class="T__StyledText-gjlic1-0 jbStwN TrackCard__TrackPrice-sc-10bk196-9 ruSDw"
              data-testid="track_price_unavailable"
            >
              Track price is unavailable
            </p>
            <p
              class="T__StyledText-gjlic1-0 jbStwN"
              data-testid="track_genre_unavailable"
            >
              Track Genere is unavailable
            </p>
          </div>
        </div>
      </div>
    </article>
  </div>
</body>
`;
